src/fuzzing/ssa.rs

```diff
@@ -37,40 +38,30 @@ pub fn validate_ssa<F: Function>(f: &F, cfginfo: &CFGInfo) -> Result<(), RegAllo
     // Walk the blocks in arbitrary order. Check, for every use, that
     // the def is either in the same block in an earlier inst, or is
     // defined (by inst or blockparam) in some other block that
-    // dominates this one. Also check that for every block param and
-    // inst def, that this is the only def.
-    let mut defined = vec![false; f.num_vregs()];
+    // dominates this one.
+    let mut local = HashSet::new();
     for block in 0..f.num_blocks() {
         let block = Block::new(block);
-        for blockparam in f.block_params(block) {
-            if defined[blockparam.vreg()] {
-                return Err(RegAllocError::SSA(*blockparam, Inst::invalid()));
-            }
-            defined[blockparam.vreg()] = true;
-        }
+        local.clear();
+        local.extend(f.block_params(block));
+
         for iix in f.block_insns(block).iter() {
             let operands = f.inst_operands(iix);
             for operand in operands {
                 match operand.kind() {
                     OperandKind::Use => {
-                        let def_block = if vreg_def_inst[operand.vreg().vreg()].is_valid() {
-                            cfginfo.insn_block[vreg_def_inst[operand.vreg().vreg()].index()]
-                        } else {
-                            vreg_def_blockparam[operand.vreg().vreg()].0
-                        };
-                        if def_block.is_invalid() {
-                            return Err(RegAllocError::SSA(operand.vreg(), iix));
-                        }
-                        if !cfginfo.dominates(def_block, block) {
+                        let def_block = defined_in[operand.vreg().vreg()];
+                        let okay = def_block.is_valid()
+                            && if def_block == block {
+                                local.contains(&operand.vreg())
+                            } else {
+                                cfginfo.dominates(def_block, block)
+                            };
+                        if !okay {
                             return Err(RegAllocError::SSA(operand.vreg(), iix));
                         }
                     }
-                    OperandKind::Def => {
-                        if defined[operand.vreg().vreg()] {
-                            return Err(RegAllocError::SSA(operand.vreg(), iix));
-                        }
-                        defined[operand.vreg().vreg()] = true;
-                    }
+                    OperandKind::Def => {}
```

Comment:
Perhaps a comment here (inside the block) saying that defs are processed below, after all uses are checked?